### ISSUE TITLE
Reading row groups

### ISSUE THREAD

**Issue by tischi** [tlabs-data/tablesaw-parquet#83]
Hi, 

Thanks a lot for developing and maintaining this super useful library!

I was wondering about reading ["row groups"](https://github.com/apache/parquet-format?tab=readme-ov-file#glossary); is that possible?

The way I understand this, this should allow one to lazy-load a subset of rows from a large data set?

**Comment by ccleva** [comment]
Hi @tischi, thanks for your kind words!

As far as I know it is not possible directly: row groups are an internal parquet structure used in [predicate pushdown](https://docs.cloudera.com/documentation/enterprise/6/6.3/topics/cdh_ig_predicate_pushdown_parquet.html#concept_pgs_plb_mgb). Predicate pushdown is not yet implemented (and there is no plan to do it in the near future).

If your use case is to read only the first N rows we could add that feature in the options, let me know if this would fit your needs.

**Comment by tischi** [comment]
Hi @ccleva,

Let me provide you with give some context:

Currently, I am not using Parquet yet, but exploring the options.

My main application, ideally, would be lazy loading from a big S3 hosted table into a Java JTable, such that data is only loaded while people move around in the JTable UI, both lazy-loading rows and columns on-demand from the object store.
It would become part of this Java application: https://www.nature.com/articles/s41592-023-01776-4
There, we already can lazy-load very well the image data, using a S3 Zarr backend, however the tables are currently stored as CSV on github an must be downloaded as a whole.

Do you know whether such lazy-loading of rows and columns from Parquet/S3 into a JTable should, in principle, be possible?

**Comment by tischi** [comment]
Thank you!

That's a bit of a bummer ;-)

I actually hoped that Parquet would "natively" support chunked loading and writing (from S3...).

I googled a bit and it seems that the python pyarrow library does support both chunked loading and saving chunks?!

https://blog.clairvoyantsoft.com/efficient-processing-of-parquet-files-in-chunks-using-pyarrow-b315cc0c62f9

For the saving, as you mentioned, there are actually now a number of files, but the library seems to help quite a bit with the creation of those chunk files, so that would be fine (for me). 

And for the loading, I would hope that `parquet_file.read_row_group(i, i + chunk_size)` would be smart enough to only touch the files that contain relevant data?

What do you think? Is it that the parquet file format in principle would support chunking, but maybe the current Java implementations don't have all the features?

**Comment by ccleva** [comment]
Yes, pyarrow supports chunking, partitioning, predicate pushdown, and reading from s3. There are some nice examples [here](https://stackoverflow.com/questions/56522977/using-predicates-to-filter-rows-from-pyarrow-parquet-parquetdataset).

Other libraries might also provide this (like [awswrangler](https://github.com/aws/aws-sdk-pandas)), but it is not a "native" feature of parquet, more a feature of the library reading and writing the files.

**Comment by tischi** [comment]
I see, and there is no Java equivalent to pyarrow...?

**Comment by ccleva** [comment]
The [java arrow implementation](https://arrow.apache.org/docs/java/) (arrows is a memory data format) includes support for [reading parquet files](https://arrow.apache.org/cookbook/java/dataset.html#query-parquet-file) but some of the features for manipulating the data afterwards are [not pure Java implementations](https://arrow.apache.org/docs/java/overview.html).

This is why I started this project: I wanted a pure Java implementation of a dataset (tablesaw is kind of a pandas equivalent) that can be used with parquet files...

**Comment by ccleva** [comment]
Moving this to the discussions
